packages/realm-react/src/cachedObject.ts

```diff
@@ -0,0 +1,83 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+import Realm from "realm";
+import { cachedCollection } from "./cachedCollection";
+
+type IndexibleRealmObject = Realm.Object & { [key: string]: unknown };
+
+export function cachedObject<T extends Realm.Object>(
+  object: T | null,
+  updateCallback: () => void,
+  listCaches = new Map(),
+  listTeardowns: (() => void)[] = [],
+): { object: (T & Realm.Object) | null; tearDown: () => void } {
+  // If the object doesn't exists, just return it with an noop tearDown
+  if (object === null) {
+    return { object, tearDown: () => undefined };
+  }
+
+  const cachedObjectHandler: ProxyHandler<T & Realm.Object> = {
+    get: function (target, key) {
+      const value = Reflect.get(target, key);
+      // Pass methods through
+      if (typeof value === "function") {
+        return value.bind(target);
+      }
+
+      // If its a Realm.List we need to add a proxy cache around it
+      if (value instanceof Realm.List) {
+        if (listCaches.has(key)) {
+          // Return in a new proxy so that the reference gets updated (otherwise it will not rerender)
```

Comment:
```suggestion
          // Return a new proxy wrapping the cachedCollection so that its reference gets updated,
          // otherwise the list component will not re-render. The cachedCollection then ensures that 
          // only the modified children of the list component actually re-render.
```